### PR TITLE
Check weak pointer to avoid potential use-after-free crash

### PR DESCRIPTION
**Pull request by 0xZOne**
Fixes: https://github.com/flutter/flutter/issues/95467

/cc @gaaclarke 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by 0xZOne** [comment]
It's a refactor that should be covered by existing tests.

File: shell/common/shell.cc [hunk 1]
```diff
@@ -1060,11 +1060,12 @@ void Shell::OnPlatformViewSetNextFrameCallback(const fml::closure& closure) {
   FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
 
   task_runners_.GetRasterTaskRunner()->PostTask(
-      [rasterizer = rasterizer_->GetWeakPtr(), closure = closure]() {
+      fml::MakeCopyable([rasterizer = rasterizer_->GetWeakPtr(),
+                         closure = std::move(closure)]() mutable {
```
Comment (chinmaygarde) [inline shell/common/shell.cc:1064]: `closure` is a const reference. So `std::move(closure)` is still going to be doing a copy like before. We have had issues in the past where there were captures in the closure that could only be destroyed on specific threads were dropped when the closure was collected on an unexpected thread. The solution there is to use a capture by pointer and then collect it on the right thread. If this patch is doing anything, it is only delaying the collection in a non-deterministic manner.

**Review by chinmaygarde — CHANGES_REQUESTED** [review]

File: shell/common/shell.cc (hunk 1 above)
Comment (0xZOne) [reply shell/common/shell.cc]: Thanks, I learned.

**Comment by Hixie** [comment]
> It's a refactor that should be covered by existing tests. @Hixie can you give a test exemption? Thanks. :)

It seems like a bug fix rather than a refactor? Refactors don't change the behaviour, just the way the behaviour is reached. This seems like it's intended to fix a bug, so has to presumably be a change in behaviour.

**Comment by chinmaygarde** [comment]
This is a sound fix that definitely makes things better. It isn't even a matter of defensive coding. All weaks must be checked for validity before use on the calling thread as a matter of course. That this wasn't done before has lead to the crash.

To test this, a contrived case we could come up with is to create the application and add its FlutterView, but ensure that the application doesn't push any frames. Then, remove the FlutterView (`surfaceUpdated:NO`) with some (not sure if platform channels will be torn down) mechanism to tell the application to start pushing frames. This is all early and late lifecycle stuff so I am not sure how convoluted it will be to do this deterministically.

Either way, I think it makes things better and the flaw and fix are easy to reason about. I leave the insistence on the test to @Hixie

**Comment by Hixie** [comment]
We should clearly commit the fix regardless. Correct code that we might regress is better than known-bad code that we might make even worse.

That said, I would encourage you to try to write a test because historically we have often found that anything that's not tested _will_ eventually end up broken again.

**Comment by chinmaygarde** [comment]
Sounds good. Landing this with [an issue to backfill the test](https://github.com/flutter/flutter/issues/96955). I feel comfortable the with this patch because checking that a weak pointer is valid should be a matter of course.

**Review by chinmaygarde — APPROVED** [review]

**Review by jason-simmons — APPROVED** [review]

**Pull request merged**